nomos-cli/src/da/network/swarm.rs

```diff
@@ -0,0 +1,136 @@
+// std
+// crates
+use kzgrs_backend::common::blob::DaBlob;
+use libp2p::futures::StreamExt;
+use libp2p::swarm::ConnectionId;
+use libp2p::Multiaddr;
+use libp2p::{identity::Keypair, swarm::SwarmEvent, PeerId, Swarm};
+use nomos_core::da::BlobId;
+use nomos_da_network_core::protocols::dispersal::executor::behaviour::{
+    DispersalError, DispersalExecutorEvent,
+};
+use nomos_da_network_core::{
+    protocols::dispersal::executor::behaviour::DispersalExecutorBehaviour, SubnetworkId,
+};
+use nomos_libp2p::{DialError, DialOpts};
+use subnetworks_assignations::MembershipHandler;
+use tokio::sync::mpsc::UnboundedSender;
+use tracing::{debug, error};
+// internal
+
+#[derive(Debug, Clone)]
+pub enum DispersalEvent {
+    /// A blob successfully arrived its destination
+    DispersalSuccess {
+        blob_id: BlobId,
+        subnetwork_id: SubnetworkId,
+    },
+    /// Something went wrong delivering the blob
+    DispersalError { error: DispersalError },
+}
+
+pub struct ExecutorSwarm<Membership>
+where
+    Membership: MembershipHandler<Id = PeerId, NetworkId = SubnetworkId> + 'static,
+{
+    swarm: Swarm<DispersalExecutorBehaviour<Membership>>,
+    dispersal_broadcast_sender: UnboundedSender<DispersalEvent>,
+}
+
+impl<Membership> ExecutorSwarm<Membership>
+where
+    Membership: MembershipHandler<NetworkId = SubnetworkId, Id = PeerId> + Clone + Send,
+{
+    pub fn new(
+        key: Keypair,
+        membership: Membership,
+        dispersal_broadcast_sender: UnboundedSender<DispersalEvent>,
+    ) -> Self {
+        let swarm = Self::build_swarm(key, membership);
+        Self {
+            swarm,
+            dispersal_broadcast_sender,
+        }
+    }
+
+    pub fn blobs_sender(&self) -> UnboundedSender<(SubnetworkId, DaBlob)> {
+        self.swarm.behaviour().blobs_sender()
+    }
+
+    fn build_swarm(
+        key: Keypair,
+        membership: Membership,
+    ) -> Swarm<DispersalExecutorBehaviour<Membership>> {
+        libp2p::SwarmBuilder::with_existing_identity(key)
+            .with_tokio()
+            .with_quic()
+            .with_behaviour(|_key| DispersalExecutorBehaviour::new(membership))
+            .expect("Validator behaviour should build")
+            .build()
+    }
+
+    pub fn dial(&mut self, addr: Multiaddr) -> Result<ConnectionId, DialError> {
+        let opt = DialOpts::from(addr.clone());
+        let connection_id = opt.connection_id();
+
+        tracing::debug!("attempting to dial {addr}. connection_id:{connection_id:?}",);
+        self.swarm.dial(opt)?;
+        Ok(connection_id)
+    }
```

Comment:
Nitpick: I think you can use the `self.swarm.dial` directly (iirc Multiaddress implements Into<DialOpts>). But im not sure.

Reply:
Used your suggestion. Will reintroduce connection id when needed.